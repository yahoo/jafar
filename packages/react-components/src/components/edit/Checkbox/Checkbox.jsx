/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import CheckboxInternal from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';

/**
 * Represent a boolean value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Checkbox/Checkbox.jsx">Checkbox</a>
 from '@jafar-org/react-components/edit/Checkbox'
 */
export default class Checkbox extends React.Component {
  static propTypes = {
    value: PropTypes.bool,
    state: PropTypes.shape({
      label: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: false,
    state: {
      label: '',
    },
    disabled: false,
  };

  render() {
    return (
      <FormControlLabel
        control={
          <CheckboxInternal
            checked={this.props.value}
            disabled={this.props.disabled}
            onChange={this.onChange}
          />
        }
        label={this.props.state.label}
      />
    );
  }

  onChange = (e, checked) => {
    this.props.onValueChange(checked);
  };
}
