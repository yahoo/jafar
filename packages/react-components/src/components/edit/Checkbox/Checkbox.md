
<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    agreeToTerms: {
      label: 'Agree to site Terms',
      path: 'agreeToTerms',
      component: {
        name: 'Checkbox',
        state: {
            label: 'I read the site terms and I agree to the terms',
        }
      },
      required: true,
      validators: [{
          name: 'mustAgree',
      }],
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      agreeToTerms: undefined,
  },
};

const resources = {
  components: { 
      Checkbox: {
          renderer: Checkbox,
      }, 
  },
  validators: {
      mustAgree: {
          func: props => props.value === true,
          message: () => 'You must agree to the terms',
      },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='agreeToTerms' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: true,
};

<Checkbox
    value={state.value}
    onValueChange={(value) => {
        setState({ value });
    }}
/>
```

<h4>With label</h4>

```javascript
initialState = { 
    value: true,
    state: {
        label: 'Click me',
    }
};

<Checkbox
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
/>
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    value: undefined,
};

<Checkbox
    value={state.value}
    onValueChange={(value) => {
        setState({ value });
    }}
/>
```

<h4>Disabled</h4>

```javascript
initialState = { 
    value: true,
};

<Checkbox
    value={state.value}
    disabled={true}
    onValueChange={(value) => {
        setState({ value });
    }}
/>
```