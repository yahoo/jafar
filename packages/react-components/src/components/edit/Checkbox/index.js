/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import Checkbox from './Checkbox';

export default Checkbox;
