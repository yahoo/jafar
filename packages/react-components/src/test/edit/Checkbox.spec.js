/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Checkbox from '../../edit/Checkbox/index';

describe('<Checkbox />', () => {
  let component;
  let onValueChangeSpy;
  const value = false;
  const state = { label: 'Click me' };
  const disabled = false;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onValueChange callback on check change', () => {
    component = mount(
      getComponent(value, state, disabled, onValueChangeSpy)
    );
    const newValue = true;
    component.find('input[type="checkbox"]').simulate('change', {
      target: { checked: newValue },
    });

    expect(onValueChangeSpy).toHaveBeenCalledWith(newValue);
  });

  function getComponent(value, state, disabled, onValueChange) {
    return (
      <Checkbox
        value={value}
        state={state}
        disabled={disabled}
        onValueChange={onValueChange}
      />
    );
  }
});
