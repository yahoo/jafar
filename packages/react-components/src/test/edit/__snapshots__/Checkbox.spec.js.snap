// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render provided data 1`] = `
<WithStyles(ForwardRef(FormControlLabel))
  control={
    <ForwardRef(WithStyles)
      checked={false}
      disabled={false}
      onChange={[Function]}
    />
  }
  label="Click me"
/>
`;
